docs/data-product-studio/data-quality/failed-events/recovering-failed-events/recover-with-sql/index.md

```diff
@@ -0,0 +1,502 @@
+    FROM SNOWPLOW_FAILED.EVENTS
+    WHERE load_tstamp::DATE > CURRENT_DATE - INTERVAL '7 DAY'
+    AND app_id = 'test-app'
+    AND contexts_com_snowplowanalytics_snowplow_failure_1[0]:schema = 'iglu:com.example/user_entity/jsonschema/2-0-1'
+  )
+SELECT
+  ARRAY_CONSTRUCT( -- since this is an entity, it must be an array
+   OBJECT_CONSTRUCT(
+   '_schema_version', '2-0-0', -- set the schema version to the correct entity version for Snowflake Streaming Loader
+   'user_id', contexts_com_snowplowanalytics_snowplow_failure_1[0]:data:user_id::INT, -- use the `[0]` syntax to extract the first (and only) instance of this entity from the failure array
+   'user_name', contexts_com_snowplowanalytics_snowplow_failure_1[0]:data:user_name::STRING  -- use the `...::INT` and the `...::STRING` functions to cast the raw JSON values to their appropriate types
+   )
+   ) AS CONTEXTS_COM_EXAMPLE_USER_ENTITY_2
+FROM prep
+```
+
+  </TabItem>
+</Tabs>
+
+## Reinserting the repaired rows
+
+Now we have successfully identified our problematic rows and corrected them, we can now insert them back to the events table. While you can hand write an `INSERT` statement for the rows, the Snowplow events tables tend to have very large and complex table schemas, with hundreds or thousands of columns, with many complex nested columns. This process is time consuming and error prone, so we recommend using the `INFORMATION_SCHEMA` tables to programmatically construct the `INSERT` command before running it.
+
+There are some complexities to bear in mind when doing this:
+
+* There may well be columns that are not in the failed events table that are in the good events table. Your insert command must include `CAST(NULL AS ... )` in place for all of those columns.
+* You must ensure that you have the correct structure for your repaired columns as otherwise they will not insert and the warehouse will reject the inserted rows.
+* In newer versions of Snowplow data warehouse loaders (such as Snowflake Streaming Loader and the BigQuery Loader V2), entity columns also have a `_schema_version` column. This must also be set for reprocessed entity columns in order for the insert to be successful.
+* If the recovered events are net-new to your good events table (say you've encountered failures on the first time using this schema) then that column won't exist in your good events table yet. We'd recommend sending a hand crafted good event for that schema to your prod pipeline **first**. This will allow the Snowplow loader to correctly create the column in your good events table so it is ready to be inserted into.
+
+:::info
+We would recommend using a staging table to insert the repaired rows before running this process on your main events table. This will allow you to review the repaired rows before inserting them into your main events table. To do so you can take a cut of your main events table, and run the `INSERT` command against that table, and verify you get the behaviour you expect.
```

Comment:
```suggestion
We would recommend using a staging table to insert the repaired rows before running this process on your main events table. This will allow you to review the repaired rows before inserting them into your main events table. To do so you can take a cut of your main events table, and run the `INSERT` command against that table, and verify you get the behavior you expect.
```